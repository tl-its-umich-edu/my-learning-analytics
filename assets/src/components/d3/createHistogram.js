import * as d3 from 'd3'
import { adjustViewport } from '../../util/chart'
import { roundToXDecimals } from '../../util/math'
import { siteTheme } from '../../globals'

function createHistogram ({
  data, width, height, domElement, xAxisLabel, yAxisLabel, myGrade, maxGrade = 100,
  showNumberOnBars = false, showDashedLine = true
}) {
  const margin = { top: 20, right: 20, bottom: 50, left: 40 }
  const barColor = siteTheme.palette.secondary.main
  const [aWidth, aHeight] = adjustViewport(width, height, margin)

  // data usually will be [50.6, 50.6, 50.6, 50.6, 50.6, 74.28, 74.52, 75.89, 76.69,,.,.,.,.] lowest grades binned
  //  to hide low performers.

  // the set operation removes duplicates
  const tempUniqData = [...new Set(data)]
  const firstGradeAfterBinnedGrade = tempUniqData[1]

  /* only showing the tick values above the binned grades eg, with above data the tick will start from 75%
  * if course has 0% or > 95% grades then we just show the whole distribution
  * */
  const minTickGrade = showDashedLine ? Math.round(firstGradeAfterBinnedGrade / 5) * 5 : 0

  const x = d3.scaleLinear()
    .domain([0, maxGrade]).nice()
    .range([margin.left, aWidth - margin.right])

  const bins = d3.histogram()
    .domain(x.domain())
    .thresholds(x.ticks(40))(data)

  // getting the first bin that has some grades in them, accessing the x1(higher bin) value
  const dashLine = () => {
    for (const bin in bins) {
      if (bins[bin].length > 0) {
        return bins[bin].x1
      }
    }
  }

  const y = d3.scaleLinear()
    .domain([0, d3.max(bins, d => d.length)]).nice()
    .range([aHeight - margin.bottom, margin.top])

  const svg = d3.select(domElement).append('svg')
    .attr('width', aWidth)
    .attr('height', aHeight)

  const bar = svg.selectAll('rect')
    .data(bins).enter()
    .append('g')

  bar.append('rect')
    .attr('x', d => x(d.x0) + 1)
    .attr('width', d => Math.max(0, x(d.x1) - x(d.x0) - 1))
    .attr('y', d => y(d.length))
    .attr('height', d => y(0) - y(d.length))
    .attr('fill', barColor)

  // Default is not show count on bar text, in case of AB testing a feature we want to leave it a configurable option
  if (showNumberOnBars) {
    bar.append('text')
      .attr('x', d => x((d.x1 + d.x0) / 2))
      .attr('y', d => y(d.length) + margin.top - 5)
      .attr('text-anchor', 'middle')
      .attr('fill', 'white')
      .text(d => d.length === 0 ? '' : d.length)
  }

  const xAxis = g => g
    .attr('transform', `translate(0, ${aHeight - margin.bottom})`)
    .call(d3
      .axisBottom(x)
      .tickSizeOuter(0)
      .ticks(width > 750 ? 20 : 10)
      .tickFormat(d => `${d}%`)
    )
    .call(g => g.append('text')
      .attr('x', aWidth / 2)
      .attr('y', 40)
      .attr('fill', 'rgba(0, 0, 0, 0.87)')
      .attr('font-size', '0.875rem')
      .attr('font-weight', '400')
      .attr('line-height', '1.46429em')
      .text(xAxisLabel).attr('dy', -4)
    )
    .call(g => g.selectAll('.tick').filter(d => d < minTickGrade).remove())

  const yAxis = g => g
    .attr('transform', `translate(${margin.left},0)`)
    .attr('class', 'axis')
    .call(d3.axisLeft(y).tickSizeInner(-aWidth).ticks(5))
    .call(g => g.select('.domain').remove())
    .call(g => g.select('.tick:last-of-type text').clone()
      .attr('x', 4)
      .attr('fill', 'rgba(0, 0, 0, 0.87)')
      .attr('text-anchor', 'end')
      .attr('font-size', '0.875rem')
      .attr('font-weight', '400')
      .attr('line-height', '1.46429em')
      .attr('opacity', 1)
      .attr('text-anchor', 'start')
      .text(yAxisLabel).attr('dy', -4))

  svg.append('g')
    .call(xAxis)

  svg.append('g')
    .call(yAxis)

  if (myGrade) {
    svg.append('line')
      .attr('transform', `translate(0, ${aHeight - margin.bottom})`)
      .attr('x1', x(myGrade))
      .attr('y1', -aHeight + 75)
      .attr('x2', x(myGrade))
      .attr('y2', 0)
      .attr('stroke', 'darkorange')
      .attr('stroke-width', '2')
    svg.append('text')
      .attr('x', myGrade < 5 ? x(myGrade) + 3 : x(myGrade) - 120)
      .attr('y', margin.top + 15)
      .text(`My Grade: ${roundToXDecimals(myGrade, 1)}%`)
      .attr('font-size', '0.875rem')
      .attr('font-weight', 'bold')
      .attr('line-height', '1.46429em')
      .attr('text-anchor', 'start')
  }
  // Dashed line to show difference b/w binned and normal distribution
  if (showDashedLine) {
    svg.append('line')
      .attr('transform', `translate(0, ${aHeight - margin.bottom})`)
      .attr('x1', x(dashLine()))
      .attr('y1', -aHeight)
      .attr('x2', x(dashLine()))
      .attr('y2', 0)
      .attr('stroke-dasharray', ('3, 3'))
      .attr('stroke', 'rgba(0, 0, 0, 0.54)')
    svg.append('text')
      .attr('x', x(dashLine()) - 40)
      .attr('y', margin.top - 5)
      .text(`<${Math.trunc(firstGradeAfterBinnedGrade)}%`)
      .attr('font-size', '0.875rem')
      .attr('line-height', '1.46429em')
      .attr('text-anchor', 'start')
      .attr('fill', 'rgba(0, 0, 0, 0.54)')
  }
}

export default createHistogram
