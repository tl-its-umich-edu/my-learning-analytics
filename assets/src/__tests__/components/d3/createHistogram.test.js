/* global describe, test, expect */
import createHistogram from '../../../components/d3/createHistogram'
import { histogramDataWithBinning } from '../../testData/d3TestData'

describe('createHistogram', () => {
  test('should build a histogram bar chart binning last five grades with MyGrade line', () => {
    const div = document.createElement('div')
    const gradeSummary = { current_user_grade: 76.8, graph_upper_limit: 100, tot_students: 84, grade_avg: 90.49, median_grade: 93.39, show_number_on_bars: true, show_dash_line: true }
    createHistogram({
      data: histogramDataWithBinning,
      width: 1000,
      height: 500,
      domElement: div,
      myGrade: 76.8,
      gradesSummary: gradeSummary
    })
    expect(div).toMatchSnapshot()
  })
})
