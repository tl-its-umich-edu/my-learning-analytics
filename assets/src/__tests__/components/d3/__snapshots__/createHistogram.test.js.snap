// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createHistogram should build a histogram bar chart binning last five grades with MyGrade line 1`] = `
<div>
  <svg
    height="430"
    width="940"
  >
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.60000000000001"
        x="41"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.599999999999994"
        x="58.60000000000001"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.59999999999998"
        x="76.2"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.600000000000023"
        x="93.79999999999998"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.599999999999994"
        x="111.4"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.599999999999994"
        x="129"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.600000000000023"
        x="146.6"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.599999999999994"
        x="164.20000000000002"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.599999999999994"
        x="181.8"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.599999999999994"
        x="199.4"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.600000000000023"
        x="217"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.599999999999966"
        x="234.60000000000002"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.600000000000023"
        x="252.2"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.600000000000023"
        x="269.8"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.599999999999966"
        x="287.40000000000003"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.600000000000023"
        x="305"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.599999999999966"
        x="322.6"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.600000000000023"
        x="340.2"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.600000000000023"
        x="357.8"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.599999999999966"
        x="375.40000000000003"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.599999999999966"
        x="393"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.600000000000023"
        x="410.59999999999997"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.60000000000008"
        x="428.2"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.59999999999991"
        x="445.80000000000007"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.600000000000023"
        x="463.4"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.600000000000023"
        x="481"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.600000000000023"
        x="498.6"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.600000000000023"
        x="516.2"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.599999999999795"
        x="533.8000000000001"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.600000000000136"
        x="551.3999999999999"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.600000000000023"
        x="569"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.600000000000023"
        x="586.6"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.600000000000023"
        x="604.2"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.59999999999991"
        x="621.8000000000001"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.600000000000023"
        x="639.4"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="16.600000000000023"
        x="657"
        y="380"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="163.63636363636365"
        width="16.59999999999991"
        x="674.6"
        y="216.36363636363635"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="98.18181818181819"
        width="16.600000000000136"
        x="692.1999999999999"
        y="281.8181818181818"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="98.18181818181819"
        width="16.59999999999991"
        x="709.8000000000001"
        y="281.8181818181818"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="130.9090909090909"
        width="16.600000000000023"
        x="727.4"
        y="249.0909090909091"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="163.63636363636365"
        width="16.600000000000023"
        x="745"
        y="216.36363636363635"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="98.18181818181819"
        width="16.59999999999991"
        x="762.6"
        y="281.8181818181818"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="261.8181818181818"
        width="16.600000000000023"
        x="780.1999999999999"
        y="118.18181818181817"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="163.63636363636365"
        width="16.600000000000023"
        x="797.8"
        y="216.36363636363635"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="261.8181818181818"
        width="16.600000000000023"
        x="815.4"
        y="118.18181818181817"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="229.0909090909091"
        width="16.600000000000136"
        x="833"
        y="150.9090909090909"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="360"
        width="16.599999999999795"
        x="850.6000000000001"
        y="20"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="327.27272727272725"
        width="16.600000000000023"
        x="868.1999999999999"
        y="52.727272727272734"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="229.0909090909091"
        width="16.600000000000023"
        x="885.8"
        y="150.9090909090909"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="65.4545454545455"
        width="16.600000000000023"
        x="903.4"
        y="314.5454545454545"
      />
    </g>
    <g>
      <rect
        fill="#2C6496"
        height="0"
        width="0"
        x="921"
        y="380"
      />
    </g>
    <g
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="middle"
      transform="translate(0, 380)"
    >
      <path
        class="domain"
        d="M40.5,0.5H920.5"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(700.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          75%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(744.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          80%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(788.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          85%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(832.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          90%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(876.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          95%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(920.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          100%
        </text>
      </g>
      <text
        dy="-4"
        fill="rgba(0, 0, 0, 0.87)"
        font-size="0.875rem"
        font-weight="400"
        line-height="1.46429em"
        x="470"
        y="40"
      />
    </g>
    <g
      class="axis"
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="end"
      transform="translate(40,0)"
    >
      <g
        class="tick"
        opacity="1"
        transform="translate(0,380.5)"
      >
        <line
          stroke="currentColor"
          x2="940"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        >
          0
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,315.0454545454545)"
      >
        <line
          stroke="currentColor"
          x2="940"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        >
          2
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,249.5909090909091)"
      >
        <line
          stroke="currentColor"
          x2="940"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        >
          4
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,184.13636363636365)"
      >
        <line
          stroke="currentColor"
          x2="940"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        >
          6
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,118.68181818181817)"
      >
        <line
          stroke="currentColor"
          x2="940"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        >
          8
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,53.227272727272734)"
      >
        <line
          stroke="currentColor"
          x2="940"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        >
          10
        </text>
        <text
          dy="-4"
          fill="rgba(0, 0, 0, 0.87)"
          font-size="0.875rem"
          font-weight="400"
          line-height="1.46429em"
          opacity="1"
          text-anchor="start"
          x="4"
        />
      </g>
    </g>
    <line
      stroke="darkorange"
      stroke-width="2"
      transform="translate(0, 380)"
      x1="715.84"
      x2="715.84"
      y1="-355"
      y2="0"
    />
    <text
      font-size="0.875rem"
      font-weight="bold"
      line-height="1.46429em"
      text-anchor="start"
      x="595.84"
      y="35"
    >
      My Grade: 76.8%
    </text>
    <line
      stroke="rgba(0, 0, 0, 0.54)"
      stroke-dasharray="3, 3"
      transform="translate(0, 380)"
      x1="691.1999999999999"
      x2="691.1999999999999"
      y1="-430"
      y2="0"
    />
    <text
      fill="rgba(0, 0, 0, 0.54)"
      font-size="0.875rem"
      line-height="1.46429em"
      text-anchor="start"
      x="651.1999999999999"
      y="15"
    >
      &lt;74%
    </text>
  </svg>
</div>
`;
